presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,578 @@
+        geometry.getEsriGeometry().queryEnvelope(envelope);
+        return envelope;
+    }
+
+    private static void verifySameSpatialReference(OGCGeometry leftGeometry, OGCGeometry rightGeometry)
+    {
+        checkArgument(leftGeometry.getEsriSpatialReference().equals(rightGeometry.getEsriSpatialReference()), "Input Geometries Spatial Reference do not match");
+    }
+
+    // centroid is arithmetic mean of the input points
+    private static Point getPointsCentroid(OGCGeometry geometry)
+    {
+        String type = geometry.geometryType();
+        int pointCount = ((MultiVertexGeometry) geometry.getEsriGeometry()).getPointCount();
+        double xSum = 0;
+        double ySum = 0;
+        for (int i = 0; i < pointCount; i++) {
+            Point point = null;
+            if (type.equals(MULTI_POINT)) {
+                MultiPoint multiPoint = (MultiPoint) geometry.getEsriGeometry();
+                point = multiPoint.getPoint(i);
+            }
+            else if (type.equals(POLYGON)) {
+                Polygon polygon = (Polygon) geometry.getEsriGeometry();
+                point = polygon.getPoint(i);
+            }
+            xSum += point.getX();
+            ySum += point.getY();
+        }
+        return new Point(xSum / pointCount, ySum / pointCount);
+    }
+
+    // centroid is weighted length of each line segment, weight in terms of line length
+    private static Point getLineCentroid(OGCGeometry geometry)
+    {
+        Polyline polyline = (Polyline) geometry.getEsriGeometry();
+        double xSum = 0;
+        double ySum = 0;
+        double weightSum = 0;
+        for (int i = 0; i < polyline.getPathCount(); i++) {
+            Point startPoint = polyline.getPoint(polyline.getPathStart(i));
+            Point endPoint = polyline.getPoint(polyline.getPathEnd(i) - 1);
+            Line line = new Line();
+            line.setStart(startPoint);
+            line.setEnd(endPoint);
+            double weight = line.calculateLength2D();
+            weightSum += weight;
+            xSum += (startPoint.getX() + endPoint.getX()) * weight / 2;
+            ySum += (startPoint.getY() + endPoint.getY()) * weight / 2;
+        }
+        return new Point(xSum / weightSum, ySum / weightSum);
+    }
+
+    // centroid is weighted length of each polygon, weight in terms of polygon area
+    private static Point getPolygonCentroid(OGCGeometry geometry)
+    {
+        OGCMultiPolygon multiPolygon = (OGCMultiPolygon) geometry;
+        double xSum = 0;
+        double ySum = 0;
+        double weightSum = 0;
+        for (int i = 0; i < multiPolygon.numGeometries(); i++) {
+            Polygon polygon = (Polygon) multiPolygon.geometryN(i).getEsriGeometry();
+            double weight = polygon.calculateArea2D();
+            weightSum += weight;
+            double polygonXSum = 0;
+            double polygonYSum = 0;
+            for (int j = 0; j < polygon.getPointCount(); j++) {
```

Comment:
Why do you compute the centroid of a polygon as a mean of its vertexes? Based on [1] and [2] the formula is different. 

[1] http://www.seas.upenn.edu/~sys502/extra_materials/Polygon%20Area%20and%20Centroid.pdf
[2] https://en.wikipedia.org/wiki/Centroid#Properties